src/Microsoft.Extensions.Configuration.AzureAppConfiguration/ConfigurationClientManager.cs

```diff
@@ -53,16 +110,49 @@ public ConfigurationClientManager(IEnumerable<Uri> endpoints, TokenCredential cr
         internal ConfigurationClientManager(IList<ConfigurationClientWrapper> clients)
         {
             _clients = clients;
-        }
-
-        public IEnumerable<ConfigurationClient> GetAvailableClients(DateTimeOffset time)
+        }
+
+#pragma warning disable CS1998 // Async method lacks 'await' operators and will run synchronously
+        public async ValueTask<IEnumerable<ConfigurationClient>> GetAvailableClients(CancellationToken cancellationToken)
+#pragma warning restore CS1998 // Async method lacks 'await' operators and will run synchronously
         {
-            return _clients.Where(client => client.BackoffEndTime <= time).Select(c => c.Client).ToList();
+            DateTimeOffset now = DateTimeOffset.UtcNow;
+
+            _ = OptimisticRefreshFallbackClients(now, cancellationToken).SwallowUnhandledOperationCanceledException();
+
+            List<ConfigurationClient> clients = new List<ConfigurationClient>(_clients.Where(c => c.BackoffEndTime <= now).Select(c => c.Client));
+
+            if (_dynamicClients != null)
+            {
+                foreach (ConfigurationClientWrapper client in _dynamicClients)
+                {
+                    if (client.BackoffEndTime <= now)
+                    {
+                        clients.Add(client.Client);
+                    }
+                }
+            }
+
+            return clients;
         }
 
-        public IEnumerable<ConfigurationClient> GetAllClients()
+#pragma warning disable CS1998 // Async method lacks 'await' operators and will run synchronously
+        public async ValueTask<IEnumerable<ConfigurationClient>> GetAllClients(CancellationToken cancellationToken)
```

Comment:
With this implementation, we don't even need ValueTask, we can return IEnumerable directly.

Reply:
True! updated